R/LSproblemofpointscommon.R

```diff
@@ -1,227 +1,234 @@
   }
-  return(p_calculated)
+  return(pCalculated)
 }
 
+
+
+# This is a recursive function returning all situations (combinations of points obtained 
+# by each player) in which player 1 wins the game.
+# Input k: a vector of containing the number of trials that each player needs to win the game.
+
 combinations <- function(k){
-  if (length(k) == 3){ # the number of players
-    #k <- c(1,2,3)
+  n <- length(k)
+  
+  if (n == 3){ # the number of players
     num <- prod(k)/k[1] #total number of sets
-    #sets <- vector(mode = "list", length = num)
-    sets <- matrix(nrow = num, ncol = length(k))
+    sets <- matrix(nrow = num, ncol = n)
     for (i in 0:(k[2]-1)){
       for (j in 0:(k[3]-1)){
-        sets[i*k[3]+j+1,] <- c(k[1],i,j)
+        sets[i*k[3]+j+1, ] <- c(k[1],i,j)
       }
     }
     return(sets)
   }else{
-    #k <- c(1,2,3,4)
-    #x1 <- sets
-    x1 <- combinations(k[1:(length(k) - 1)]) # k without k[length(k)]
-    x2 <- k[length(k)]
-    y <- matrix(ncol = length(k), nrow = prod(k)/k[1])
-    for (i in 1:length(x1[,1])){
-      for (j in 1:k[length(k)]){
-        y[(i-1)*(k[length(k)])+j,] <- append(x1[i,],j-1)
+    x1 <- combinations(k[1:(n-1)]) # recursion
+    x2 <- k[n]
+    sets <- matrix(ncol = n, nrow = prod(k)/k[1])
+    for (i in 1:length(x1[ ,1])){
+      for (j in 1:k[n]){
+        sets[(i-1)*(k[n])+j, ] <- append(x1[i, ],j-1)
       }
     }
-    return(y)
+    
+    return(sets)
   }
 }
 
-compare_function1 <- function(p,m1,n1,t,simulation){
+compareChanceTwoPlayers <- function(p,m1,n1,t,simulation){
 
   # first estimating the probability that player 1 wins by simulating the game for a given number of times.
   m <- t - m1 # m: the number of success that player 1 should have to win the game
   n <- t - n1 # n: the number of success that player 2 should have to win the game
 
-  record_game <- vector()
+  recordGame <- vector()
   for (i in 1:simulation){
     # copy m and n
-    record_trial <- rbind(m, n)
+    recordTrial <- rbind(m, n)
     # keep running the game until one of them wins
-    while (record_trial[1, ] * record_trial[2, ] != 0){
-      record_trial <- record_trial - rmultinom(1, 1, c(p, 1 - p)) #updating the recorded result
+    while (recordTrial[1, ] * recordTrial[2, ] != 0){
+      recordTrial <- recordTrial - rmultinom(1, 1, c(p, 1 - p)) #updating the recorded result
     }
 
-    # if record_trial[1,] = 0, record 1,player 1 wins this single trial;
-    # if record_trial[2,] = 0, record 0, player 2 wins
-    record_game[i] <- record_trial[2, ]^record_trial[1, ]
+    # if recordTrial[1, ] = 0, record 1,player 1 wins this single trial;
+    # if recordTrial[2, ] = 0, record 0, player 2 wins
+    recordGame[i] <- recordTrial[2, ]^recordTrial[1, ]
 
   }
-  p_simulated <- sum(record_game)/simulation    # estimated probability that player 1 wins
+  pSimulated <- sum(recordGame)/simulation    # estimated probability that player 1 wins
```

Comment:
So we have
```r
while (recordTrial[1, ] * recordTrial[2, ] != 0) {
  recordTrial <- recordTrial - rmultinom(1, 1, c(p, 1 - p))
}
```
which implies that after the loop either `recordTrial[1, ] == 0` or `recordTrial[2, ] == 0`. The other value can be any integer right?

Slightly more readable than `recordGame[i] <- recordTrial[2, ]^recordTrial[1, ]` would be
`recordGame[i] <- if (recordTrial[2, ] == 0) 0 else 1`.

Also, isn't `pSimulated <- sum(recordGame)/simulation` equal to `pSimulated <- mean(recordGame)`?

Reply:
Yes, this makes it much more readable.